text/0000-improved-ember-registry-apis.md

```diff
@@ -0,0 +1,540 @@
+
+Methods which support either `Identifier`s or `FactoryTypeIdentifier`s should display the appropriate variant of the message according to what the user actually supplied.
+
+## How we teach this
+
+- **New concepts:** this RFC introduces the ideas of `Identifier` and `FactoryTypeIdentifier` to define the new identifier types passed as arguments to the various registry functions. The basic shapes of these objects, and perhaps their names (as inline text like “identifier” and “factory type identifier”) will need to be integrated into the guides and API docs. See also [**Unresolved Questions**](#unresolved-questions): there may be other names which work better.
+
+- The existing documentation about dependency injection can largely remain unchanged—mostly, it can be simplified in certain key areas. Paragraphs explaining the microsyntax can be eliminated or simplified by showing the identifier object type. New Ember users will continue to benefit from the detailed introduction to the concepts in the guides, but will have one fewer concept to learn along the way.
+
+- All instances of the API documentation will need to be updated to reflect the changes to their signatures.
+
+- The deprecation message must be added to Ember’s deprecation page and tooling.
+
+- For the duration of Ember 3.x releases, the Guides and API docs will need explanations of the previous, deprecated APIs. These explanations should be brief, focused on how to migrate away from them preferably via the supplied codemod, rather than diving into the details of how the legacy system worked.
+
+## Drawbacks
+
+- The existing APIs have served Ember well for a long time. Introducing any new change involves some risk, particularly in areas so core to how Ember behaves.
+
+- The proposed API here trades away brevity in favor of clarity, tooling support, and being “normal” JavaScript. `'service:session'` is many fewer characters to type than `{ type: 'service', name: 'session' }`.
+
+## Alternatives
+
+### A new String-based API
+
+Instead of providing an object-based API, we could provide a new string-based API. The `lookup` method, then, would have a signature like this:
+
+```ts
+interface Owner {
+  lookup(type: string, name: string, options: Options);
+}
+```
+
+This has the advantage of being a smaller diff from today’s world, and cutting down on the extra typing of the new API design:
+
+```diff
+- lookup('service:foo')
++ lookup('service', 'foo')
+```
+
+vs.
+
+```diff
+- lookup('service:foo');
++ lookup({ type: 'service', name: 'foo' })
+```
+
+This is a less flexible proposal, and—more importantly—it has a significant downside for usability beyond the simplest case. Supporting namespaces here would require *another* string argument, and functions with multiple arguments of the same type are notoriously easy to misuse. (“Wait, what’s the order? Namespace first or type first? Does namespace come after type, or after name?” etc.)
+
+### String-based API as “sugar”
```

Comment:
I question whether this is even worth mentioning as a possible additional API given the problems outlined in the previous "New string based API" section. Given that this RFC's general thrust seems to be "trading away sugar for explicitness and unlocking better tooling possibilities", it almost feels weird to even put the option of keeping some of the sugar around into the conversation.

Reply:
I broadly agree, but figured I'd raise it explicitly because it *will* come up in the RFC conversation proper. Better to call it out, I think?

Reply:
I broadly agree, but figured I'd raise it explicitly because it *will* come up in the RFC conversation proper. Better to call it out, I think?

Reply:
Ya, agree that it seems good to include. But I ultimately hope we decide not to implement it this way...

Reply:
Concur! I added some notes on why the RFC recommends against this, hopefully capturing both of your points; see b4b07a6.